### PR TITLE
feat(scripts): parse scripts block

### PR DESCRIPTION
**Pull request by wmalik**
# Reason for This Change

This PR is a building block for the upcoming `scripts` functionality.

## Description of Changes

* Parse all `script` blocks into  `config.Scripts`
* Perform basic validation of a `script` block
  - labels of a `script` block
  - `script.description` attribute
  - `script.job` blocks
  - mutual exclusion check between `script.job.command` and `script.job.commands`  
* Parse a `script` block only when `terramate.config.experiments` contains `"scripts"`
### Script examples

```hcl
script "terraform" "deploy" {
  description = "Terraform Deploy"
  job {
    commands = [
      ["terraform", "init"],
      ["terraform", "plan"],
      ["terraform", "apply", "-auto-approve"],
  }
}
```

```hcl
script "groupname" "scriptname" {
  description = "description of the command"
  job {
    command = ["ls", "-l", "-r"]
  }
}
```

```hcl
script "groupname" "scriptname" {
  description = "description of the command"
  job {
    command = ["ls", "-l", "-r"]
  }
  job {
    command = ["stat", "."]
  }
}
```

NOTE: functions/globals in scripts will be implemented in the next PR

### CODE REVIEW

**Comment by soerenmartius** [comment]
Cool stuff! I am wondering if we should come up with a black list of commands (e.g. `rm -rf /` :p)

File: hcl/hcl_script.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+			errs.Append(err)
+		}
+	}
+
+	if err := errs.AsError(); err != nil {
+		return nil, err
+	}
+
+	parsed := Commands(cmdsAttr)
+	return &parsed, nil
```
Comment (i4ki) [inline hcl/hcl_script.go:209]: same as for `validateDescription()`. We cannot evaluate them at this time.

**Review by i4ki — CHANGES_REQUESTED** [review]
It's good overall but the evaluation of attributes must not be done at the parser time, as it could contain interpolated variables that will only be available later.
Other than that, also some adjusts in the error messages are required, in order to give the correct line numbers and better hints of the error.

File: hcl/hcl_script.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+func (p *TerramateParser) parseScriptBlock(block *ast.Block) (*Script, error) {
+	errs := errors.L()
+
+	if len(block.Labels) != 2 {
+		errs.Append(errors.E(ErrScriptTwoLabels, block.OpenBraceRange))
+	} else if block.Labels[0] == "" {
+		errs.Append(errors.E(ErrScriptEmptyLabel, block.OpenBraceRange))
+	}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+	parsedScript := &Script{
+		Labels: block.Labels,
+	}
+
+	if len(block.Attributes) == 0 {
+		errs.Append(errors.E(ErrScriptNoAttrs, block.OpenBraceRange))
+	}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+		switch attr.Name {
+		case "description":
+			desc, err := p.validateDescription(attr)
+			if err != nil {
+				errs.Append(errors.E(ErrScriptInvalidDesc, attr.NameRange))
+				continue
+			}
+			parsedScript.Description = desc
+		default:
+			errs.Append(errors.E(ErrScriptUnknownAttr, attr.NameRange))
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+		}
+	}
+
+	if len(block.Blocks) < 1 {
+		errs.Append(errors.E(ErrScriptNoBlocks, block.TypeRange))
+	}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+			parsedScript.Jobs = append(parsedScript.Jobs, parsedJobBlock)
+		default:
+			errs.Append(errors.E(ErrScriptUnknownBlock, nestedBlock.TypeRange, nestedBlock.Type))
+
+		}
+	}
+
+	if err := errs.AsError(); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Good idea, fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+	if len(block.Attributes) == 0 {
+		errs.Append(errors.E(ErrScriptNoAttrs, block.OpenBraceRange))
+	}
+
+	for _, attr := range block.Attributes {
+		switch attr.Name {
+		case "description":
+			desc, err := p.validateDescription(attr)
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+		return ast.Attribute{}, errs.AsError()
+	}
+
+	if val.AsString() == "" {
+		errs.Append(errors.E("empty description"))
+		return ast.Attribute{}, errs.AsError()
+	}
+
+	return attr, nil
+}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+func validateScriptJobBlock(block *ast.Block) (*ScriptJob, error) {
+	errs := errors.L()
+
+	var foundCmd, foundCmds bool
+	parsedScriptJob := &ScriptJob{}
+	for _, attr := range block.Attributes {
+		switch attr.Name {
+		case "command":
+			cmdAttr, err := validateCommand(attr)
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

File: hcl/hcl_script.go (hunk 1 above)
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/8e6f17a968d313064e25a3be6d9d0e715f6e34be

**Comment by wmalik** [comment]
> Cool stuff! I am wondering if we should come up with a black list of commands (e.g. `rm -rf /`, `sudo`, etc :p)

Good point, it could make sense to allow users to provide a command blacklist, such that if a command ever evaluates to an item in that blacklist it should be rejected. I took a note of that, and will see if we can add it when we build the `terramate run-script` functionality.

File: hcl/hcl_script.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+		}
+	}
+
+	if len(parsedScript.Labels) == 0 {
+		errs.Append(errors.E(ErrScriptNoLabels, block.TypeRange))
+	}
+
+	if len(parsedScript.Jobs) == 0 {
+		errs.Append(errors.E(ErrScriptMissingOrInvalidJob, block.OpenBraceRange))
```
Comment (i4ki) [inline hcl/hcl_script.go:98]: I think `block.Range` must be used here so it's clear that the `script` body is missing a `job` definition.

File: hcl/hcl_script.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// job.command and job.commands are mutually exclusive
+	if parsedScriptJob.Command != nil && parsedScriptJob.Commands != nil {
+		errs.Append(errors.E(ErrScriptCmdConflict, parsedScriptJob.Command.NameRange))
+		errs.Append(errors.E(ErrScriptCmdConflict, parsedScriptJob.Commands.NameRange))
+	}
+
+	if parsedScriptJob.Command == nil && parsedScriptJob.Commands == nil {
+		errs.Append(errors.E(ErrScriptNoCmds, block.TypeRange))
```
Comment (i4ki) [inline hcl/hcl_script.go:140]:
```suggestion
		errs.Append(errors.E(ErrScriptNoCmds, block.Range))
```

File: hcl/hcl_script_test.go [hunk 12]
```diff
@@ -0,0 +1,659 @@
...
+									Command: makeCommand(t, `["echo", "hello"]`),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "script with an unrecogznized child block of job",
```
Comment (i4ki) [inline hcl/hcl_script_test.go:276]:
```suggestion
			name: "script with an unrecognized child block of job",
```

**Review by i4ki — APPROVED** [review]
LGTM !!

Well done @wmalik 

I left some minor improvement ideas, it's up to you.

File: hcl/hcl.go [hunk 13]
```diff
@@ -1406,7 +1408,12 @@ func parseVendorConfig(cfg *VendorConfig, vendor *ast.Block) error {
 	return errs.AsError()
 }
 
-func parseRootConfig(cfg *RootConfig, block *ast.MergedBlock) error {
+// hasExperimentalFeature returns true if the config has an experimental feature enabled
```
[the remark this answers is not in the document]
Comment (wmalik) [reply hcl/hcl.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/0bd145bc88f507782997876f06b14dc3a983a24c

File: hcl/hcl_script.go (hunk 10 above)
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/0bd145bc88f507782997876f06b14dc3a983a24c

File: hcl/hcl_script.go (hunk 11 above)
Comment (wmalik) [reply hcl/hcl_script.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/0bd145bc88f507782997876f06b14dc3a983a24c

File: hcl/hcl_script_test.go (hunk 12 above)
Comment (wmalik) [reply hcl/hcl_script_test.go]: Fixed in https://github.com/terramate-io/terramate/pull/1236/commits/0bd145bc88f507782997876f06b14dc3a983a24c

**Comment by i4ki** [comment]
@wmalik go for it !!

**Review by i4ki — APPROVED** [review]
LGTM

**Pull request merged**